File: 9_1.py
import random

# Глобальная переменная: количество элементов в массивах
ARRAY_SIZE = 10

def create_array(size):
    """Функция для создания массива случайных чисел.
    Формальный параметр: size. Фактический параметр: ARRAY_SIZE.
    Возвращаемое значение — локальная переменная.
    """
    return [random.randint(0, 20) for _ in range(size)]

def replace_max_with_zero(arr):
    """Заменяет максимальный элемент массива на 0.
    Формальный параметр: arr. Параметр передается по ссылке.
    """
    max_value = max(arr)  # Локальная переменная
    max_index = arr.index(max_value)  # Локальная переменная
    arr[max_index] = 0  # Изменение массива по ссылке
    return arr

def multiply_after_min(arr):
    """Умножает элементы массива после минимального на 2.
    Формальный параметр: arr. Параметр передается по ссылке.
    """
    min_value = min(arr)  # Локальная переменная
    min_index = arr.index(min_value)  # Локальная переменная
    for i in range(min_index + 1, len(arr)):
        arr[i] *= 2  # Изменение массива по ссылке
    return arr

# Основная программа
array1 = create_array(ARRAY_SIZE)  # Фактический параметр передается в функцию
array2 = create_array(ARRAY_SIZE)

print("Исходный массив 1:", array1)
print("Исходный массив 2:", array2)

# Проверка на наличие кратных пяти элементов и замена максимального на 0
if any(x % 5 == 0 for x in array1):
    array1 = replace_max_with_zero(array1)

# Умножение элементов после минимального во втором массиве
array2 = multiply_after_min(array2)

print("Измененный массив 1:", array1)
print("Измененный массив 2:", array2)

File: 9_1.pas
program arraymanipulation;

uses crt;

const
  array_size = 10; // глобальная константа для размера массивов

var
  array1, array2: array[1..array_size] of integer; // глобальные переменные для массивов

// процедура для заполнения массива случайными числами
procedure fillarray(var arr: array of integer; size: integer); // формальные параметры: arr, size
var
  i: integer;
begin
  randomize; // инициализация генератора случайных чисел
  for i := 0 to size - 1 do
    arr[i] := random(100); // заполнение массива случайными числами от 0 до 99
end;

// процедура для замены максимального элемента на ноль
procedure replacemaxwithzero(var arr: array of integer; size: integer); 
var
  i, maxindex: integer;
  hasmultipleoffive: boolean;
begin
  hasmultipleoffive := false;
  maxindex := 0;

  // проверяем наличие кратного пяти элемента
  for i := 0 to size - 1 do
  begin
    if (arr[i] mod 5 = 0) then
    begin
      hasmultipleoffive := true;
      break; 
    end;
  end;

  // если кратный пяти элемент найден, ищем максимальный элемент
  if hasmultipleoffive then
  begin
    for i := 0 to size - 1 do
    begin
      if (arr[i] > arr[maxindex]) then
        maxindex := i;
    end;
    arr[maxindex] := 0; // заменяем максимальный элемент на ноль
  end;
end;

// процедура для умножения элементов после минимального на 2
procedure multiplyaftermin(var arr: array of integer; size: integer); 
var
  i, minindex: integer;
begin
  minindex := 0;

  // находим индекс минимального элемента
  for i := 1 to size - 1 do
  begin
    if (arr[i] < arr[minindex]) then
      minindex := i;
  end;

  // умножаем все элементы после минимального на 2
  for i := minindex + 1 to size - 1 do
    arr[i] := arr[i] * 2;
end;

// процедура для вывода массива на экран
procedure printarray(arr: array of integer; size: integer); // формальные параметры: arr, size
var
  i: integer;
begin
  for i := 0 to size - 1 do
    write(arr[i], ' '); // выводим элементы массива
  writeln; // переход на новую строку
end;

begin
  // заполняем массивы случайными числами
  fillarray(array1, array_size); // фактические параметры: array1, array_size
  fillarray(array2, array_size); // фактические параметры: array2, array_size

  writeln('исходный массив 1:');
  printarray(array1, array_size);
  
  writeln('исходный массив 2:');
  printarray(array2, array_size);

  // выполняем операции с массивами
  replacemaxwithzero(array1, array_size); // передача параметров по ссылке
  multiplyaftermin(array2, array_size);    // передача параметров по ссылке

  writeln('массив 1 после замены максимального элемента:');
  printarray(array1, array_size);

  writeln('массив 2 после умножения элементов:');
  printarray(array2, array_size);
end.
